### PR TITLE
Event listeners for decision engine

### PR DESCRIPTION
**Pull request by mramendi**

### CODE REVIEW

File: doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-con.adoc [hunk 1]
```diff
@@ -2,35 +2,53 @@
 
 = {DECISION_ENGINE_CAP} event listeners and debug logging
 
-In {PRODUCT}, you can add or remove listeners for {DECISION_ENGINE} events, such as fact insertions and rule executions. With {DECISION_ENGINE} event listeners, you can be notified of {DECISION_ENGINE} activity and separate your logging and auditing work from the core of your application.
+The {DECISION_ENGINE} generates _events_ when performing activities such as fact insertions and rule executions. If you register _event listeners_, the {DECISION_ENGINE} calls every every listener when an activity is performed.
```
Comment (mariofusco) [inline doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-con.adoc:5]: Duplicate word "every".

**Review by mariofusco — APPROVED** [review]

File: doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-con.adoc (hunk 1 above)
Comment (hmanwani-rh) [inline doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-con.adoc:5]:
```suggestion
The {DECISION_ENGINE} generates events when performing activities such as fact insertions and rule executions. If you register event listeners, the {DECISION_ENGINE} calls every listener when an activity is performed.
```

File: doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-con.adoc [hunk 2]
```diff
@@ -2,35 +2,53 @@
...
 
-The {DECISION_ENGINE} supports the following default event listeners for the agenda and working memory:
+Event listeners have methods that correspond to different types of activities. The {DECISION_ENGINE} passes an event object to each method; this object contains information about the specific activity.
 
-* `AgendaEventListener`
-* `WorkingMemoryEventListener`
+Your code can implement custom event listeners. It can also add and remove registered event listeners. In this way, your code can be notified of {DECISION_ENGINE} activity, and you can separate logging and auditing work from the core of your application.
```
Comment (hmanwani-rh) [inline doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-con.adoc:9]:
```suggestion
Your code can implement custom event listeners and you can also add and remove registered event listeners. In this way, your code can be notified of {DECISION_ENGINE} activity, and you can separate logging and auditing work from the core of your application.
```

File: doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-con.adoc [hunk 3]
```diff
@@ -2,35 +2,53 @@
...
-* `ObjectDeletedEvent`
-* `ObjectUpdatedEvent`
-* `ProcessCompletedEvent`
-* `ProcessNodeLeftEvent`
-* `ProcessNodeTriggeredEvent`
-* `ProcessStartEvent`
+{PRODUCT} includes default implementations of these listeners: `DefaultAgendaEventListener` and `DefaultRuleRuntimeEventListener`. You can extend each of these implementations to monitor specific events.
 
-For example, the following code uses a `DefaultAgendaEventListener` listener attached to a KIE session and specifies the `AfterMatchFiredEvent` event to be monitored. The code prints pattern matches after the rules are executed (fired):
+For example, the following code extends `DefaultAgendaEventListener` to monitor the  `AfterMatchFiredEvent` event and attaches this listener to a KIE session. The code prints pattern matches when rules are executed (fired):
```
Comment (hmanwani-rh) [inline doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-con.adoc:51]:
```suggestion
For example, the following code extends `DefaultAgendaEventListener` to monitor the  `AfterMatchFiredEvent` event and attaches this listener to a KIE session. The code prints a pattern that matches when rules are executed (fired):
```

File: doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-development-con.adoc [hunk 4]
```diff
@@ -0,0 +1,12 @@
+[id='engine-event-listeners-development-con_{context}']
+= Practices for development of event listeners
+
+The {DECISION_ENGINE} calls event listeners during rule processing. The calls block the execution of the {DECISION_ENGINE}. Therefore, the event listener can affect the  performance of the {DECISION_ENGINE}.
```
Comment (hmanwani-rh) [inline doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-development-con.adoc:4]:
```suggestion
The {DECISION_ENGINE} calls event listeners during rule processing. The calls block the execution of the {DECISION_ENGINE}. Therefore, the event listener can affect the performance of the {DECISION_ENGINE}.
```

File: doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-development-con.adoc [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+To ensure minimal disruption, observe the following practices:
```
Comment (hmanwani-rh) [inline doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-development-con.adoc:6]:
```suggestion
To ensure minimal disruption, consider the following practices:
```

File: doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-development-con.adoc [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+* Any action must be as short as possible.
+* A listener class must not have a state. The {DECISION_ENGINE} can destroy and re-create a listener class at any time.
+* Do not use logic that relies on the order of execution of different event listeners.
+* Do not include interaction with different entities outside the {DECISION_ENGINE} within a listener. For example, do not include REST calls for notification of events. An excepting is output of logging information; however, a logging listener must be as simple as possible.
```
Comment (hmanwani-rh) [inline doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-development-con.adoc:11]:
```suggestion
* Do not include interaction with different entities outside the {DECISION_ENGINE} within a listener. For example, do not include REST calls for notification of events. An excepting is an output of logging information; however, a logging listener must be as simple as possible.
```

**Review by hmanwani-rh — COMMENTED** [review]
@mramendi A few minor comments are added, rest looks good. 
thanks for taking up this task.

File: doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-development-con.adoc (hunk 5 above)
Comment (emmurphy1) [reply doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-development-con.adoc:6]: I would say:
To ensure minimal disruption, observe the following guidelines:
or
To ensure minimal disruption, follow these guidelines:

**Review by hellowdan — APPROVED** [review]

**Pull request merged**